cnxdb/archive-sql/schema/triggers.sql

```diff
@@ -248,54 +248,67 @@ CREATE TRIGGER ruleset_trigger
 
 -- recipe maintenace triggers
 
-CREATE OR REPLACE FUNCTION update_default_recipes() RETURNS trigger AS '
+CREATE OR REPLACE FUNCTION update_default_recipes() RETURNS trigger AS $$
 BEGIN
-  IF (TG_OP = ''INSERT'' OR TG_OP = ''UPDATE'') AND
+  IF (TG_OP = 'INSERT' OR TG_OP = 'UPDATE') AND
           NEW.revised >= ((SELECT revised FROM print_style_recipes
               WHERE print_style = NEW.print_style ORDER BY revised DESC LIMIT 1)
-              UNION ALL VALUES (NEW.revised) LIMIT 1)
+              UNION ALL VALUES (NEW.revised) LIMIT 1) -- avoid compare to NULL
   THEN
       DELETE FROM default_print_style_recipes WHERE print_style = NEW.print_style;
       INSERT into default_print_style_recipes (
-        print_style, tag, fileid, recipe_type, revised)
+        print_style, title, tag, commit_id, fileid, recipe_type, revised)
   	VALUES (
-        NEW.print_style, NEW.tag, NEW.fileid, NEW.recipe_type, NEW.revised);
+        NEW.print_style, NEW.title, NEW.tag, NEW.commit_id, NEW.fileid, NEW.recipe_type, NEW.revised);
   END IF;
 
-  IF TG_OP = ''UPDATE'' THEN
-      UPDATE default_print_style_recipes SET
-        fileid=NEW.fileid,
-        recipe_type=NEW.recipe_type,
-        NEW.revised)
-        WHERE print_style=NEW.print_style AND tag=NEW.tag;
+  IF TG_OP = 'UPDATE' THEN
+    PERFORM 1 from modules where recipe = OLD.fileid;
+    IF FOUND and NEW.fileid != OLD.fileid THEN
+        RAISE WARNING 'Cannot change recipe file: recipe in use';
+        RETURN NULL;
+    END IF;
+    UPDATE default_print_style_recipes SET
+      tag=NEW.tag,
+      title=NEW.title,
+      fileid=NEW.fileid,
+      recipe_type=NEW.recipe_type
+      WHERE print_style=NEW.print_style and revised=NEW.revised;
   END IF;
 
 RETURN NEW;
 END;
 
-' LANGUAGE 'plpgsql';
+$$ LANGUAGE 'plpgsql';
 
 CREATE TRIGGER update_default_recipes
   BEFORE INSERT OR UPDATE ON print_style_recipes FOR EACH ROW
   EXECUTE PROCEDURE update_default_recipes();
 
 
-CREATE OR REPLACE FUNCTION delete_from_default_recipes() RETURNS trigger AS '
+CREATE OR REPLACE FUNCTION delete_from_default_recipes() RETURNS trigger AS $$
```

Comment:
Can you add something like a doc string to these trigger functions? Looks like this function has the side-effect of inserting a new recipe.

Reply:
done